### PR TITLE
Simple Merkle Tree

### PR DESCRIPTION
**Pull request by laskdaf**
- updated checkMembership() and tests to support tendermint's simple merkle tree
- simple merkle proofs require the total number of txns in the tree, so i added a parameter in submitBlock() that takes in a list of the roots' txn count.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -6,8 +6,8 @@ before_script:
   - npm install -g truffle@4.1.14 ganache-cli@6.1.8
   - npm install
 
-script: 
-  - ganache-cli -m=plasma_mvp > /dev/null &
+script:
+  - ganache-cli -m "they only media any modify banner suffer pole tag rule creek harvest" > /dev/null &
```
Comment (AdityaSripal) [inline .travis.yml]: what is this?

File: contracts/libraries/Validator.sol [hunk 2]
```diff
@@ -5,33 +5,74 @@ import "openzeppelin-solidity/contracts/ECRecovery.sol";
 library Validator {
     uint8 constant WORD_SIZE = 32;
 
+    // temporary placeholder function so rootchain will compile
+    function checkMembership(bytes32 leaf, uint256 index, bytes32 rootHash, bytes proof)
```
Comment (AdityaSripal) [inline contracts/libraries/Validator.sol]: placeholder should be removed and real function called `checkMembership`

File: contracts/rootchain/RootChain.sol [hunk 3]
```diff
@@ -104,8 +107,8 @@ contract RootChain is Ownable {
                 root := mload(add(memPtr, i))
```
Comment (AdityaSripal) [inline contracts/rootchain/RootChain.sol]: Think it makes more sense to have:

```
numBlocks = numTxns.length()
for (uint i = 0; i < numBlocks; i++)
```

Then mload with 32 * i

File: contracts/rootchain/RootChain.sol [hunk 4]
```diff
@@ -104,8 +107,8 @@ contract RootChain is Ownable {
...
             }
 
-            childChain[currentChildBlock] = childBlock(root, block.timestamp);
-            emit BlockSubmitted(root, currentChildBlock);
+            childChain[currentChildBlock] = childBlock(root, numTxns[i/32], block.timestamp);
```
Comment (AdityaSripal) [inline contracts/rootchain/RootChain.sol]: Then turn this to `i` with comment above

File: contracts/rootchain/RootChain.sol [hunk 5]
```diff
@@ -171,7 +174,13 @@ contract RootChain is Ownable {
         require(txList.length == 17, "incorrect number of items in the transaction list");
 
         sigList = spendMsg[1].toList();
-        require(sigList.length == 2, "two signatures must be present");
+        // require(sigList.length == 2, "two signatures must be present"); // FIXME add this line back later
```
Comment (AdityaSripal) [inline contracts/rootchain/RootChain.sol]: this needs to be added back in?

File: test/rootchain/rootchain_helpers.js [hunk 6]
```diff
@@ -23,74 +24,130 @@ let fastForward = async function(time) {
...
 
-// For a given list of leaves, this function generates a merkle root. It assumes
-// the merkle tree is of depth 16. If there are less than 2^16 leaves, the
-// list is padded with 0x0 transactions. The function also generates a merkle
-// proof for the leaf at txIndex.
-// @param leaves The leaves for which this function generates a merkle root and proof
-// @param txIndex The leaf for which this function gneerates a merkle proof
-let generateMerkleRootAndProof = function(leaves, txIndex) {
-    return generateMerkleRootAndProofHelper(leaves, 16, txIndex, 0);
+// // For a given list of leaves, this function generates a merkle root. It assumes
```
Comment (AdityaSripal) [inline test/rootchain/rootchain_helpers.js]: why is all this code commented out? needs cleanup

File: test/rootchain/rootchain_helpers.js [hunk 7]
```diff
@@ -23,74 +24,130 @@ let fastForward = async function(time) {
...
-                proof = leaves[i + 1].slice(2);
-            }
-            else if (txIndex == i + 1) {
-                proof = leaves[i].slice(2);
+let sha256StringMultiple = function(input1, input2) {
+    let toHash = "0x20" + input1.slice(2) + "20" + input2.slice(2);
+    return toHex(ethjs_util.sha256(toHash).toString('hex'));
+};
+
+let generateMerkleRootAndProofNew = function(leaves, index) {
```
Comment (AdityaSripal) [inline test/rootchain/rootchain_helpers.js]: This needs to replace old function

File: .travis.yml (hunk 1 above)
Comment (laskdaf) [reply .travis.yml]: so i included tests that verify the inputs you gave me. i.e. the challenges. so i need to hard code in the sender's private key. 

if you don't think these types of tests are necessary, i'll remove them

Comment (hamdiallam) [reply .travis.yml]: do we have to use the hardcoded values?

File: contracts/rootchain/RootChain.sol [hunk 8]
```diff
@@ -203,14 +208,14 @@ contract RootChain is Ownable {
...
-        require(txHash.checkSigs(keccak256(abi.encodePacked(merkleHash, blk.root)), // confirmation hash -- sha3(merkleHash, root)
+        bytes32 merkleHash = sha256(txBytes);
+        require(txHash.checkSigs(sha256(abi.encodePacked(merkleHash, blk.root)), // confirmation hash -- sha256(merkleHash, root)
                          // we always assume the first input is always present in a transaction. The second input is optional
                          txList[6].toUint() > 0 || txList[9].toUint() > 0, // existence of input1. Either a deposit or utxo
                          sigList[0].toBytes(), sigList[1].toBytes(), confirmSignatures), "signature mismatch");
 
         // check proof
-        require(merkleHash.checkMembership(txPos[1], blk.root, proof), "invalid merkle proof");
+        require(merkleHash.checkMembership(txPos[1], blk.root, proof, blk.numTxns), "invalid merkle proof"); // FIXME: should not hardcode total=2
```
Comment (legengliu) [inline contracts/rootchain/RootChain.sol]: could you remove this FIXME? it was left over from the last time i worked on this.

File: contracts/rootchain/RootChain.sol [hunk 9]
```diff
@@ -203,14 +208,14 @@ contract RootChain is Ownable {
         childBlock storage blk = childChain[txPos[0]];
 
         // check signatures
-        bytes32 merkleHash = keccak256(txBytes);
-        require(txHash.checkSigs(keccak256(abi.encodePacked(merkleHash, blk.root)), // confirmation hash -- sha3(merkleHash, root)
+        bytes32 merkleHash = sha256(txBytes);
+        require(txHash.checkSigs(sha256(abi.encodePacked(merkleHash, blk.root)), // confirmation hash -- sha256(merkleHash, root)
```
Comment (legengliu) [inline contracts/rootchain/RootChain.sol]: the variable `txHash` is returned by the `decodeTransaction` helper, which creates `txHash` with a keccak hash: `txHash = keccak256(spendMsg[0].toRlpBytes());` (RootChain.sol line 182). Should that be changed to a sha256 hash as well?

File: contracts/libraries/Validator.sol [hunk 10]
```diff
@@ -8,30 +8,64 @@ library Validator {
...
+    function computeHashFromAunts(uint256 index, uint256 total, bytes32 leaf, bytes innerHashes)
+        internal
+        pure
+        returns (bytes32)
+    {
+        require(index < total, "Index must be less than total number of leaf nodes");
+        require(total > 0, "Must have at least one leaf node");
+
+        if (total == 1) {
+            require(innerHashes.length == 0, "computeHashFromAunts 1");
```
Comment (AdityaSripal) [inline contracts/libraries/Validator.sol]: This error message means nothing

**Pull request merged**